### PR TITLE
feat: add time based paymaster

### PR DESCRIPTION
**Pull request by dutterbutter**
Changes introduced in this PR:
- Adds time based paymaster example

### CODE REVIEW

File: contracts/contracts/paymasters/TimeBasedPaymaster.sol [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    );
+
+    bytes4 paymasterInputSelector = bytes4(_transaction.paymasterInput[0:4]);
+    if (paymasterInputSelector == IPaymasterFlow.general.selector) {
+      uint256 startTime = (block.timestamp / 86400) * 86400 + 14 hours;
+      uint256 endTime = startTime + 10 minutes;
+
+      require(
+        block.timestamp >= startTime && block.timestamp <= endTime,
+        "Transactions can only be processed between 9:00 AM - 9:03 AM EST"
```
Comment (idea404) [inline contracts/contracts/paymasters/TimeBasedPaymaster.sol:46]: Should this time gap not be 10 minutes considering 10min is being added to the start time variable above?

Comment (idea404) [reply contracts/contracts/paymasters/TimeBasedPaymaster.sol:46]: Also the times might be incorrect too, I believe it is 14:00 to 14:10?

File: contracts/test/timeBased.test.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    currentDate.setUTCMinutes(1);
+    currentDate.setUTCSeconds(0);
+    currentDate.setUTCMilliseconds(0);
+    const targetTime = Math.floor(currentDate.getTime() / 1000);
+    await provider.send("evm_setNextBlockTimestamp", [targetTime]);
+
+    // Act
+    const initialBalance = await userWallet.getBalance();
+    await executeGreetingTransaction(userWallet);
+    await provider.send("evm_mine", []);
```
Comment (idea404) [inline contracts/test/timeBased.test.ts:71]: Would make it more complete imo to also check the greeting has been set/changed

File: contracts/contracts/paymasters/TimeBasedPaymaster.sol (hunk 1 above)
Comment (dutterbutter) [reply contracts/contracts/paymasters/TimeBasedPaymaster.sol:46]: The time is correct, 14:00 is UTC time. I will update nonetheless to be more universal.

**Review by idea404 — APPROVED** [review]

**Pull request merged**